src/adr/0051-pulsar-mailqueue.md

```diff
@@ -0,0 +1,197 @@
+
+## Technical details
+
+[[This section requires a deep review]]
+
+[Akka](https://akka.io/) actor system is used in single node mode as a processing framework.
+
+The MailQueue relies on the following topology:
+
+ - out topic : emails are first enqueued here.
+ - scheduled topic: emails that are delayed are first enqueued there.
+ - filter topic: Deletions (name, sender, recipients) prior a given sequence are synchronized between nodes using this topic.
+ 
+[[Question]]
+```
+https://github.com/apache/james-project/pull/808#discussion_r780181972
+
+Likely a stupid question but why using a separate topic for delayed messages?
+
+My take is that it limits the out-of-order impact and yields better performance.
+
+Do you confirm?
+```
+
+Upon enqueue, the blobs are first saved, then the Pulsar message payload is generated and published to the relevant 
+topic (out or scheduled).
+
+Scheduled messages have their `deliveredAt` property set to the desired value. When the delay is 
+expired, the message will be consumed and thus moved to the out topic. Flushes simply copy content of the scheduled
+topic to the out topic then reset the offset of the scheduled queue, atomically. Expired filters are removed.
+
+[[Question]]
+```
+https://github.com/apache/james-project/pull/808#discussion_r780170243
+
+> Expired filters are removed.
+
+Given we could get emails with arbitrary delays, invalidating filters likely lead to bugs, as the underlying emails can 
+be out of order.
+
+Maybe it is safer to get rid of this optimisation. Correct?
+```
+
+The size of the mail queue can be simply computed from the out and scheduled topics.
+
+Upon deletes, the condition of this deletion, as well as the sequence before which it applies is synchronized across
+nodes an in-memory datastructures wrapped in an actor. Each instance uses a unique subscription and thus will maintain a
+set of all deletions ever performed.
+
+Upon dequeues, messages of the out topic are filtered using that in-memory data structure, then exposed as a reactive 
+publisher.
+
+Upon browsing, both the out and scheduled topic are read from the consumption offset and filtering is applied.
+
+Upon clear, the out topic is deleted.
+
+[[Question]]
+```
+https://github.com/apache/james-project/pull/808/files#r780162174
+
+Shouldn't clear also get rid of the scheduled topic?
+
+After clearing, can messages still transit through the mailqueue? (I expect it should, but I do not see how this can 
+work if we nuke the out topic... New producers would re-create automatically a missing topic?)
+```
+
+Miscellaneous remarks:
+
+ - The pulsar admin client is used to list existing queues.
+ - Priorities are not yet supported.
+ - Only metadata transit through Pulsar. The general purpose James blobStore, backed by a S3 compatible API, is used to
+   store the underlying email content. Saves on top of object-storage is latency prone and exposed to end SMTP clients.
```

Comment:
is latency a problem in async handling?

Reply:
enqueue is synchronous

(dequeue is asynchronous)

Reply:
Yes, we could even change enqueue to be reactive, I can't see any reason not too

Reply:
We already have a reactive enqueue, the problem lies down to the fact that SMTP implementation is blocking (that's complex to change).

Here I am not arguing against the blobStore abstraction nor am I arguing against a distributed one. I am arguing that some blobStore implementation (like S3 backed with OVH) have a performance profile poorly suited for a queue usage. THe requirements are just very different from long term storage. While long term storage could accomodate higher write latencies (written asynchronously) in order to get cheaper storage costs (think HDD) on a queue it makes more sense to have fast but expensive storage (think SSD). The trade offs are not the same. The optimal technologies to implement the blobStore in both case might not be the same either... And this is an orthogonal problem to reactive enqueues.